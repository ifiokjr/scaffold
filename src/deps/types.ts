export * from "https://cdn.skypack.dev/type-fest@2.13.1?dts";
