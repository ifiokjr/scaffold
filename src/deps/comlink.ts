export * from "https://esm.sh/comlink@4?target=deno";
