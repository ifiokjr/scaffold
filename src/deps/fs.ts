export * from "https://deno.land/std@0.144.0/fs/mod.ts";
export * from "https://esm.sh/merge-anything?target=deno";
