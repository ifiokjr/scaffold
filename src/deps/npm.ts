export { render } from "https://deno.land/x/eta@v1.12.3/mod.ts";

export * from "https://esm.sh/is-what@4.1.7?target=deno";
export * from "https://esm.sh/merge-anything@5.0.2?target=deno";
export { default as invariant } from "https://esm.sh/tiny-invariant@1.2.0?target=deno";
export * from "https://esm.sh/ts-extras@0.10.2?target=deno";
