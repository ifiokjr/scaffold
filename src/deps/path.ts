export { default as normalizePath } from "https://cdn.skypack.dev/normalize-path@v3.0.0?dts";
export * as path from "https://deno.land/std@0.144.0/path/mod.ts";
export * from "https://esm.sh/junk@4.0.0?type=deno";
