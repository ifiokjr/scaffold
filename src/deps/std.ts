export { Tar, Untar } from "https://deno.land/std@0.145.0/archive/tar.ts";
export { parse } from "https://deno.land/std@0.145.0/flags/mod.ts";
export { readLines } from "https://deno.land/std@0.145.0/io/mod.ts";
export {
  BaseHandler,
  type HandlerOptions,
} from "https://deno.land/std@0.145.0/log/handlers.ts";
export {
  type LevelName,
  LogLevels,
} from "https://deno.land/std@0.145.0/log/levels.ts";
export { Logger, LogRecord } from "https://deno.land/std@0.145.0/log/logger.ts";
export {
  globToRegExp,
  isGlob,
} from "https://deno.land/std@0.145.0/path/glob.ts";
export {
  copy,
  iterateReader,
  readableStreamFromIterable,
  readableStreamFromReader,
  readerFromIterable,
  readerFromStreamReader,
} from "https://deno.land/std@0.145.0/streams/conversion.ts";
export {
  assert,
  equal,
} from "https://deno.land/std@0.145.0/testing/asserts.ts";
