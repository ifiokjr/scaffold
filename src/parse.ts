import { LiteralUnion } from "./deps/types.ts";
import { matchOne } from "./utils/match.ts";

export interface GitRepository {
  site: string;
  user: string | undefined;
  name: string | undefined;
  ref: LiteralUnion<"HEAD", string>;
  url: string;
  ssh: string;
  subdirectory: string | undefined;
  mode: "tar" | "git";
}
const GIT_URL_REGEX =
  /^(?:(?:https:\/\/)?(?<urlSite>[^/:]+\.[^/:]+)\/|git@(?<gitSite>[^/:]+)[/:]|(?<namedSite>[^/]+):)?(?<user>[^\s/]+)\/(?<name>[^\s#/]+)(?:(?<subdirectory>(?:\/[^\s#/]+)+))?\/?(?:#(?<ref>.+))?/;

/**
 * From the provided string, extract the hostname and the path.
 *
 * Adapted from https://github.com/Rich-Harris/degit/blob/64b80577acf3313b669840f7452800ee8d09fbf3/src/index.js#L327-L364
 */
export function parseGitUrl(source: string): GitRepository | undefined {
  const match = matchOne(source, GIT_URL_REGEX);

  if (!match) {
    return;
  }

  const { unnamed } = match;

  const site = (unnamed[0] || unnamed[1] || unnamed[2] || "github").replace(
    /\.(com|org)$/,
    "",
  );

  if (!SUPPORTED.has(site)) {
    return;
  }

  const user = unnamed[3];
  const name = unnamed[4]?.replace(/\.git$/, "");
  const subdirectory = unnamed[5];
  const ref = unnamed[6] || "HEAD";

  const domain = `${site}.${
    site === "bitbucket" ? "org" : site === "git.sr.ht" ? "" : "com"
  }`;
  const url = `https://${domain}/${user}/${name}`;
  const ssh = `git@${domain}:${user}/${name}`;
  const mode = SUPPORTED.has(site) ? "tar" : "git";

  return { site, user, name, ref, url, ssh, subdirectory, mode };
}

const SUPPORTED = new Set(["github", "gitlab", "bitbucket", "git.sr.ht"]);
