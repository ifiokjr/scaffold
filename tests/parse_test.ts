import { parseGitUrl } from "../src/parse.ts";
import { assertSnapshot, describe, it } from "./deps.ts";

describe("parseGitUrl", () => {
  // Tests taken from https://github.com/Rich-Harris/degit/blob/64b80577acf3313b669840f7452800ee8d09fbf3/test/test.js#L43-L121
  const urls = {
    github: [
      "ifiokjr/scaffold#main",
      "test_org/test_repo",
      "github:test_org/test_repo",
      "git@github.com:test_org/test_repo",
      "https://github.com/test_org/test_repo.git",
    ],
    gitlab: [
      "gitlab:Rich-Harris/degit-test-repo",
      "git@gitlab.com:Rich-Harris/degit-test-repo",
      "https://gitlab.com/Rich-Harris/degit-test-repo.git",
    ],
    bitbucket: [
      "bitbucket:Rich_Harris/degit-test-repo",
      "git@bitbucket.org:Rich_Harris/degit-test-repo",
      "https://bitbucket.org/Rich_Harris/degit-test-repo.git",
    ],
    sourcehut: [
      "git.sr.ht/~satotake/degit-test-repo",
      "https://git.sr.ht/~satotake/degit-test-repo",
      "git@git.sr.ht:~satotake/degit-test-repo",
      "Rich-Harris/degit-test-repo/subdir",
    ],
    subdirectories: [
      "test_org/scaffold/subdir#v2.0.0",
      "Rich-Harris/degit-test-repo/subdir",
      "github:Rich-Harris/degit-test-repo/subdir",
      "git@github.com:Rich-Harris/degit-test-repo/subdir",
      "https://github.com/Rich-Harris/degit-test-repo.git/subdir",
    ],
  };

  for (const [name, namedUrls] of Object.entries(urls)) {
    for (const url of namedUrls) {
      it(`should parse ${name} url: ${url}`, async (t) => {
        await assertSnapshot(t, parseGitUrl(url));
      });
    }
  }
});
