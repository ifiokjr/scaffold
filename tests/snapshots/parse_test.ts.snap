export const snapshot = {};

snapshot[`parseGitUrl > should parse github url: ifiokjr/scaffold#main 1`] = `
{
  mode: "tar",
  name: "scaffold",
  ref: "main",
  site: "github",
  ssh: "git@github.com:ifiokjr/scaffold",
  subdirectory: undefined,
  url: "https://github.com/ifiokjr/scaffold",
  user: "ifiokjr",
}
`;

snapshot[`parseGitUrl > should parse github url: test_org/test_repo 1`] = `
{
  mode: "tar",
  name: "test_repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:test_org/test_repo",
  subdirectory: undefined,
  url: "https://github.com/test_org/test_repo",
  user: "test_org",
}
`;

snapshot[`parseGitUrl > should parse github url: github:test_org/test_repo 1`] = `
{
  mode: "tar",
  name: "test_repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:test_org/test_repo",
  subdirectory: undefined,
  url: "https://github.com/test_org/test_repo",
  user: "test_org",
}
`;

snapshot[`parseGitUrl > should parse github url: git@github.com:test_org/test_repo 1`] = `
{
  mode: "tar",
  name: "test_repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:test_org/test_repo",
  subdirectory: undefined,
  url: "https://github.com/test_org/test_repo",
  user: "test_org",
}
`;

snapshot[`parseGitUrl > should parse github url: https://github.com/test_org/test_repo.git 1`] = `
{
  mode: "tar",
  name: "test_repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:test_org/test_repo",
  subdirectory: undefined,
  url: "https://github.com/test_org/test_repo",
  user: "test_org",
}
`;

snapshot[`parseGitUrl > should parse gitlab url: gitlab:ifiokjr/scaffold-test 1`] = `
{
  mode: "tar",
  name: "scaffold-test",
  ref: "HEAD",
  site: "gitlab",
  ssh: "git@gitlab.com:ifiokjr/scaffold-test",
  subdirectory: undefined,
  url: "https://gitlab.com/ifiokjr/scaffold-test",
  user: "ifiokjr",
}
`;

snapshot[`parseGitUrl > should parse gitlab url: git@gitlab.com:ifiokjr/scaffold-test 1`] = `
{
  mode: "tar",
  name: "scaffold-test",
  ref: "HEAD",
  site: "gitlab",
  ssh: "git@gitlab.com:ifiokjr/scaffold-test",
  subdirectory: undefined,
  url: "https://gitlab.com/ifiokjr/scaffold-test",
  user: "ifiokjr",
}
`;

snapshot[`parseGitUrl > should parse gitlab url: https://gitlab.com/ifiokjr/scaffold-test.git 1`] = `
{
  mode: "tar",
  name: "scaffold-test",
  ref: "HEAD",
  site: "gitlab",
  ssh: "git@gitlab.com:ifiokjr/scaffold-test",
  subdirectory: undefined,
  url: "https://gitlab.com/ifiokjr/scaffold-test",
  user: "ifiokjr",
}
`;

snapshot[`parseGitUrl > should parse bitbucket url: bitbucket:Rich_Harris/degit-test-repo 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "bitbucket",
  ssh: "git@bitbucket.org:Rich_Harris/degit-test-repo",
  subdirectory: undefined,
  url: "https://bitbucket.org/Rich_Harris/degit-test-repo",
  user: "Rich_Harris",
}
`;

snapshot[`parseGitUrl > should parse bitbucket url: git@bitbucket.org:Rich_Harris/degit-test-repo 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "bitbucket",
  ssh: "git@bitbucket.org:Rich_Harris/degit-test-repo",
  subdirectory: undefined,
  url: "https://bitbucket.org/Rich_Harris/degit-test-repo",
  user: "Rich_Harris",
}
`;

snapshot[`parseGitUrl > should parse bitbucket url: https://bitbucket.org/Rich_Harris/degit-test-repo.git 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "bitbucket",
  ssh: "git@bitbucket.org:Rich_Harris/degit-test-repo",
  subdirectory: undefined,
  url: "https://bitbucket.org/Rich_Harris/degit-test-repo",
  user: "Rich_Harris",
}
`;

snapshot[`parseGitUrl > should parse sourcehut url: git.sr.ht/~satotake/degit-test-repo 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "git.sr.ht",
  ssh: "git@git.sr.ht.:~satotake/degit-test-repo",
  subdirectory: undefined,
  url: "https://git.sr.ht./~satotake/degit-test-repo",
  user: "~satotake",
}
`;

snapshot[`parseGitUrl > should parse sourcehut url: https://git.sr.ht/~satotake/degit-test-repo 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "git.sr.ht",
  ssh: "git@git.sr.ht.:~satotake/degit-test-repo",
  subdirectory: undefined,
  url: "https://git.sr.ht./~satotake/degit-test-repo",
  user: "~satotake",
}
`;

snapshot[`parseGitUrl > should parse sourcehut url: git@git.sr.ht:~satotake/degit-test-repo 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "git.sr.ht",
  ssh: "git@git.sr.ht.:~satotake/degit-test-repo",
  subdirectory: undefined,
  url: "https://git.sr.ht./~satotake/degit-test-repo",
  user: "~satotake",
}
`;

snapshot[`parseGitUrl > should parse sourcehut url: Rich-Harris/degit-test-repo/subdir 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:Rich-Harris/degit-test-repo",
  subdirectory: "/subdir",
  url: "https://github.com/Rich-Harris/degit-test-repo",
  user: "Rich-Harris",
}
`;

snapshot[`parseGitUrl > should parse subdirectories url: test_org/scaffold/subdir#v2.0.0 1`] = `
{
  mode: "tar",
  name: "scaffold",
  ref: "v2.0.0",
  site: "github",
  ssh: "git@github.com:test_org/scaffold",
  subdirectory: "/subdir",
  url: "https://github.com/test_org/scaffold",
  user: "test_org",
}
`;

snapshot[`parseGitUrl > should parse subdirectories url: Rich-Harris/degit-test-repo/subdir 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:Rich-Harris/degit-test-repo",
  subdirectory: "/subdir",
  url: "https://github.com/Rich-Harris/degit-test-repo",
  user: "Rich-Harris",
}
`;

snapshot[`parseGitUrl > should parse subdirectories url: github:Rich-Harris/degit-test-repo/subdir 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:Rich-Harris/degit-test-repo",
  subdirectory: "/subdir",
  url: "https://github.com/Rich-Harris/degit-test-repo",
  user: "Rich-Harris",
}
`;

snapshot[`parseGitUrl > should parse subdirectories url: git@github.com:Rich-Harris/degit-test-repo/subdir 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:Rich-Harris/degit-test-repo",
  subdirectory: "/subdir",
  url: "https://github.com/Rich-Harris/degit-test-repo",
  user: "Rich-Harris",
}
`;

snapshot[`parseGitUrl > should parse subdirectories url: https://github.com/Rich-Harris/degit-test-repo.git/subdir 1`] = `
{
  mode: "tar",
  name: "degit-test-repo",
  ref: "HEAD",
  site: "github",
  ssh: "git@github.com:Rich-Harris/degit-test-repo",
  subdirectory: "/subdir",
  url: "https://github.com/Rich-Harris/degit-test-repo",
  user: "Rich-Harris",
}
`;
